migrate/sqlschema/state.go

```diff
@@ -0,0 +1,386 @@
+package sqlschema
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/uptrace/bun/schema"
+)
+
+type State struct {
```

Comment:
```go
type State interface {
    Tables() []Table
}

type Table interface {
   Columns() []Column
}

table.(*pgdialect.Table)
table.(*mysqldialect.Table)
```

Reply:
After some experimenting with what these interfaces might be I arrived at a slightly different solution. 
Here's a short example:

```go
// sqlschema
type Table struct {
  Name  string
  Columns  []Column

  // Additional stores opaque metadata specific to each dialect
  Additional interface{}
}

// pgdialect
type Table struct {
  InheritsFrom  sqlschema.Table // Postgres supports table-inheritance
} 

table := Table{..., InheritsFrom: parentTable}

// elsewhere
pgtable := table.Additional.(*pgdialect.Table)
```

This achieves the same flexibility -- different dialects or even implementations of `Inspector` may extend the table with additional information and type-cast it later. That way we wouldn't need to define interfaces for what is essentially a DTO.

WDYT?

Reply:
>different dialects or even implementations of Inspector may extend the table with additional information and type-cast it later.

It works indeed but looks awkward to me.

This looks much more straightforward and achieves the same:

```go
// sqlschema

type BaseTable struct {
  Name    string
  Columns []Column
}

func (t *BaseTable) GetName() { return t.Name }
func (t *BaseTable) GetColumns() { return t.Columns }

type Table interface {
  GetName() string
  GetColumns() []Column
}

// pgdialect

type Table struct {
  sqlschema.BaseTable

  InheritsFrom  *Table // Postgres supports table-inheritance
} 

// usage

var table sqlschema.Table
table = &pgschema.Table{
  BaseTable: sqlschema.BaseTable{Name: ..., Columns: ...},
  InheritsFrom: parentTable,
}}

// elsewhere
pgtable := table.(*pgdialect.Table)
pgtable.InheritsFrom
```

But we can definitely start with you design if you insist.